File: kawazoe/q2029/tyama_codeiq2029.pas
{$apptype console} //Delphi
program StepUpSum;
var n,r,i,x:int64;
begin
	while(true) do begin
		if(eof(input)) then break;
		readln(n);
		if(n=0) then break;
		r:=0;
		i:=2;
		while(true) do begin
			x:=n-(i*(i-1) div 2);
			if(x<i) then break;
			if(x mod i=0) then r:=r+(x div i);
			i:=i+1;
		end;
		writeln(r);
	end;
end.

File: kawazoe/q2029/tyama_codeiq2029.hs
#!/usr/bin/env runghc
import Control.Applicative
import System.IO (isEOF)

calc r i n = do
	let x = n-(div (i*(i-1)) 2)
	if x>=i then do
		if (mod x i)==0 then
			calc (r+(div x i)) (i+1) n
		else
			calc r (i+1) n
	else
		r

main = do
	eof <- isEOF
	if not eof then do
		n <- readLn
		if n>0 then do
			putStrLn $ show $ calc 0 2 n
			main
		else return()
	else return ()